designs/sam-app-config.md

```diff
@@ -0,0 +1,201 @@
+SAM CLI App Level Config
+====================================
+
+
+What is the problem?
+--------------------
+
+Today users of SAM CLI need to invoke the CLI directly with all parameters supplied to its commands.
+
+for e.g: `sam build --use-container --debug`
+
+But often, during the lifecycle of building and deploying a serverless application. the same commands get run repeatedly to build, package and deploy, before solidifying into the final application. 
+
+These CLI commands are often long and have many changing parts.
+
+Have a look at the following series of workflows
+
+
+* `sam build --use-container --template ... --parameter-overrides=... --skip-pull-image --manifest ...`
+
+* `sam package --s3-bucket ... --template-file ... --output-template-file ... --s3-prefix ... --kms-key-id ...`
+
+* `sam deploy deploy --template-file ... --stack-name ... --capabilities ... --tags ... --parameter-overrides ... --kms-key-id ...`
+
+If this could be condensed into a series of workflows that look like
+
+* `sam build`
+* `sam package`
+* `sam deploy`
+
+That would be a huge user experience win.
+
+What will be changed?
+---------------------
+
+The suite of commands supported by SAM CLI would be aided by looking for a configuration file thats locally located to the template.yaml by default. 
```

Comment:
How does this work with build then? We have (at least try to) trained customers to operate on their source. In cases of build, we generate a new template for them (that they shouldn't care about). This means in some cases, we would pick the config in the source but others (that have run build before) we will be looking at the built template. Lets be strong on where this should live, otherwise this can be a point of confusion for customers.

Reply:
Moved to project level with an option to set the location of the configuration file with a environment variable.